sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -202,14 +201,15 @@ object JoinReorderDP extends PredicateHelper {
   }
 
   /**
-   * Builds a new JoinPlan if the two given sides can be joined with some conditions.
+   * Builds a new JoinPlan if there exists at least one join condition involving references from
+   * both left and right.
    * @param oneJoinPlan One side JoinPlan for building a new JoinPlan.
    * @param otherJoinPlan The other side JoinPlan for building a new join node.
```

Comment:
left and right sides are decided inside this method. It tends to build a left deep tree.